app/lib/account/google_oauth2.dart

```diff
@@ -21,19 +26,72 @@ final _logger = Logger('pub.account.google_auth2');
 /// The token-info end-point.
 final _tokenInfoEndPoint = Uri.parse('https://oauth2.googleapis.com/tokeninfo');
 
-/// Provides OAuth2-based authentication through Google accounts.
-class GoogleOauth2AuthProvider extends AuthProvider {
+/// Provides OAuth2-based authentication through JWKS and Google account APIs.
+class DefaultAuthProvider extends AuthProvider {
   late http.Client _httpClient;
   late oauth2_v2.Oauth2Api _oauthApi;
 
-  GoogleOauth2AuthProvider() {
+  DefaultAuthProvider() {
     _httpClient = http.Client();
     _oauthApi = oauth2_v2.Oauth2Api(_httpClient);
   }
 
+  @override
+  Future<JsonWebToken?> tryAuthenticateAsServiceToken(String token) async {
+    if (!JsonWebToken.looksLikeJWT(token)) {
+      return null;
+    }
+    final idToken = JsonWebToken.tryParse(token);
+    if (idToken == null) {
+      return null;
+    }
+
+    if (idToken.payload.iss == GitHubJwtPayload.issuerUrl) {
+      // At this point we have confirmed that the token is a JWT token
+      // issued by GitHub. If there is an issue with the token, the
+      // authentication should fail without any fallback.
+      await _verifyToken(idToken, openIdDataFetch: fetchGithubOpenIdData);
+    }
+
+    if (idToken.payload.iss == GcpServiceAccountJwtPayload.issuerUrl &&
+        idToken.payload.aud.length == 1 &&
+        idToken.payload.aud.single ==
+            activeConfiguration.externalServiceAudience) {
+      // As the uploader token's audience and the admin token's issuer and also
+      // their audience is the same, we only parse it as a non-user token, when
+      // the authentication source is from the pub client app (e.g. uploading a
+      // new package).
+      // At this point we don't fall back to authenticating the token as a user.
+
+      // TODO: use the tokeninfo endpoint instead
+      await _verifyToken(idToken, openIdDataFetch: fetchGoogleCloudOpenIdData);
+    }
+    return null;
+  }
+
+  Future<void> _verifyToken(
+    JsonWebToken idToken, {
+    required Future<OpenIdData> Function() openIdDataFetch,
+  }) async {
+    if (!idToken.payload.isTimely(threshold: Duration(minutes: 2))) {
+      throw AuthenticationException.tokenInvalid('invalid timestamps');
+    }
+    final aud =
+        idToken.payload.aud.length == 1 ? idToken.payload.aud.single : null;
+    if (aud != activeConfiguration.externalServiceAudience) {
+      throw AuthenticationException.tokenInvalid(
+          'audience "${idToken.payload.aud}" does not match "${activeConfiguration.externalServiceAudience}"');
+    }
+    final openIdData = await openIdDataFetch();
+    final signatureMatches = await idToken.verifySignature(openIdData.jwks);
+    if (!signatureMatches) {
+      throw AuthenticationException.tokenInvalid('invalid signature');
+    }
+  }
+
   /// Authenticate [token] as `access_token` or `id_token`.
   @override
-  Future<AuthResult?> tryAuthenticate(String token) async {
+  Future<AuthResult?> tryAuthenticateAsUser(String token) async {
```

Comment:
Can't we merge `tryAuthenticateAsServiceToken` and `tryAuthenticateAsUser` into `tryAuthenticate` ?

Reply:
`AuthResult` can just have two constructors, with some optional parameters... or we can have two subclasses of `AuthResult`.

Reply:
As discussed offline, this will be done in a follow-up.